Userland/Games/Chess/ChessWidget.cpp

```diff
@@ -538,89 +539,234 @@ ErrorOr<String> ChessWidget::get_fen() const
 
 ErrorOr<void> ChessWidget::import_pgn(Core::File& file)
 {
-    m_board = Chess::Board();
-
-    ByteBuffer bytes = TRY(file.read_until_eof());
-    StringView content = bytes;
-    auto lines = content.lines();
-    StringView line;
-    size_t i = 0;
-
-    // Tag Pair Section
-    // FIXME: Parse these tags when they become relevant
-    do {
-        line = lines.at(i++);
-    } while (!line.is_empty() || i >= lines.size());
+    reset();
+    enum TokenType {
+        MOVE,
+        TAG_SYMBOL,
+        TAG_STRING,
+        COMMENT,
+        GAME_TERMINATOR
+    };
 
-    // Movetext Section
-    bool skip = false;
-    bool recursive_annotation = false;
-    bool future_expansion = false;
-    Chess::Color turn = Chess::Color::White;
-    String movetext;
+    struct Token {
+        TokenType type;
+        StringView value;
+    };
 
-    for (size_t j = i; j < lines.size(); j++)
-        movetext = TRY(String::formatted("{}{}", movetext, lines.at(i)));
+    ByteString bytes = ByteString { TRY(file.read_until_eof()).bytes() };
+    Vector<Token> tokens;
+    StringView continuation_characters = "_+#=:-"sv;
+    size_t i = 0;
+    size_t token_start = 0;
+    size_t token_end = 0;
+    size_t line_number = 1;
+
+    // FIXME: Suffixes ! and ? have not been implemented.
+    while (i < bytes.length()) {
+        auto c = bytes[i];
+        switch (c) {
+        // Parse the tags.
+        case '[':
+            // Skip the [ character.
+            c = bytes[++i];
+            token_start = i;
+            while (c != '"' && !is_ascii_space(c)) {
+                if (i == bytes.length()) {
+                    GUI::MessageBox::show(window(), StringView { TRY(String::formatted("Reached end of file trying to parse tag on line {}, expected opening quote \"", line_number)).bytes() }, "Import Error"sv, GUI::MessageBox::Type::Information);
```

Comment:
A couple of things here:
- String has `.bytes_as_string_view()` that should be used instead of manually creating a StringView from `.bytes()`
- We're trying to move away from TRY-ing small string allocations, but in this case it's not a big deal as we're already returning ErrorOr in this function.
- A trickier thing: This function returns ErrorOr, but when we get a parsing error, we display a messagebox and then return {} so to outside code it looks like there wasn't an error. I think it'd be nicer if we returned an error containing this information instead, and the MessageBox was shown by the code that calls import_pgn(). I say this is tricky because `AK::Error` can't hold a String, only a StringView, so to include the nice message you'd need to make your own Error class, return `ErrorOr<void, YourErrorClass>`, and manually handle a couple of TRYs that happen here to convert them into that error type. Feel free to not do this if it's confusing!

Reply:
> String has .bytes_as_string_view() that should be used instead of manually creating a StringView from .bytes()

I tried that at first, but I got a compilation error that `bytes_as_string_view()` is a deleted function, so ended up doing this way just to get it working.

> I think it'd be nicer if we returned an error containing this information instead, and the MessageBox was shown by the code that calls import_pgn()

I can give this a try. As an aside, I was thinking while writing this code that it belongs in `AK::LibChess` somewhere instead of in the application so that it can be tested more readily, as there are some edgecases that currently I'm testing by hand in an ad-hoc mannar.

> We do also have AK::GenericLexer which can make this kind of parsing code a bit easier to write

I looked at this and it looks like just what I need for this! I'll add a `FIXME` about rewriting that I can tackle later.